### PR TITLE
DOC-5694 Prepared Statements

### PR DESCRIPTION
**Pull request by RichardSmedley**

### CODE REVIEW

File: modules/concept-docs/pages/n1ql-query.adoc [hunk 1]
```diff
@@ -14,6 +14,47 @@ Below, we fill in some of the gaps between reference and rolling-up-your-sleeves
...
+The computation for the plan adds some additional processing time and overhead for the query.
+
+Often-used queries can be _prepared_ so that its _plan_ is generated only once.
+Subsequent queries using the same query string will use the pre-generated plan instead, saving on the overhead and processing of the plan each time.
+
+You can indicate to the SDK that a given query should be optimized in the above fashion.
+When an SDK sees that a query should be optimized, it will internally prepare the statement and store the plan in an internal cache.
+When issuing the query again, the SDK will check to see if a plan exists in its cache, and will send the plan to the server.
+
+To indicate that an SDK should optimize a query, the `Adhoc` parameter should be set to false.
```
Comment (chvck) [inline modules/concept-docs/pages/n1ql-query.adoc]: I feel like this paragraph is slightly misleading, it defaults to false so queries use prepared statements by default. The wording here makes me think that I have to disable something to use prepared statements when in fact you have to enable AdHoc to not use them.

File: modules/concept-docs/pages/n1ql-query.adoc [hunk 2]
```diff
@@ -14,6 +14,47 @@ Below, we fill in some of the gaps between reference and rolling-up-your-sleeves
...
+When a query is not ad hoc, the SDK will fetch the plan (if it does not already have it).
+Do not turn off the `Adhoc` flag for each query since only a finite number of query plans (currently 5000) can be stored in the SDK.
```
Comment (chvck) [inline modules/concept-docs/pages/n1ql-query.adoc]: I'm pretty sure we don't actually have any hard limit on plans.

**Review by chvck — CHANGES_REQUESTED** [review]

File: modules/concept-docs/pages/n1ql-query.adoc (hunk 2 above)
Comment (RichardSmedley) [reply modules/concept-docs/pages/n1ql-query.adoc]: Thanks @chvck - will dig into this.

File: modules/concept-docs/pages/n1ql-query.adoc (hunk 1 above)
Comment (RichardSmedley) [reply modules/concept-docs/pages/n1ql-query.adoc]: Ah, is it specced for all of the SDKs like this?

File: modules/concept-docs/pages/n1ql-query.adoc (hunk 2 above)
Comment (RichardSmedley) [reply modules/concept-docs/pages/n1ql-query.adoc]: `private static final int QUERY_CACHE_SIZE = 5000;`
in Java SDK 3.0,
but I'm changing the text in any case  :+1:

**Review by chvck — APPROVED** [review]

**Pull request merged**